File: packages/csharp/HtmlToMarkdown/Metadata/TextDirection.cs
using System.Text.Json.Serialization;
using HtmlToMarkdown.Serialization;

namespace HtmlToMarkdown.Metadata;

/// <summary>
/// Represents text directionality of document content.
/// Corresponds to the HTML `dir` attribute and `bdi` element directionality.
/// </summary>
[JsonConverter(typeof(JsonPropertyNameEnumConverter<TextDirection>))]
public enum TextDirection
{
    /// <summary>
    /// Left-to-right text flow (default for Latin scripts).
    /// </summary>
    [JsonPropertyName("ltr")]
    LeftToRight,

    /// <summary>
    /// Right-to-left text flow (Hebrew, Arabic, Urdu, etc.).
    /// </summary>
    [JsonPropertyName("rtl")]
    RightToLeft,

    /// <summary>
    /// Automatic directionality detection.
    /// </summary>
    [JsonPropertyName("auto")]
    Auto,
}

File: packages/java/src/main/java/io/github/goldziher/htmltomarkdown/metadata/TextDirection.java
package io.github.goldziher.htmltomarkdown.metadata;

/**
 * Text directionality of document content.
 *
 * <p>Corresponds to the HTML {@code dir} attribute and {@code bdi} element directionality.
 *
 * @since 2.13.0
 */
public enum TextDirection {
  /** Left-to-right text flow (default for Latin scripts). */
  LEFT_TO_RIGHT("ltr"),

  /** Right-to-left text flow (Hebrew, Arabic, Urdu, etc.). */
  RIGHT_TO_LEFT("rtl"),

  /** Automatic directionality detection. */
  AUTO("auto");

  private final String value;

  TextDirection(String value) {
    this.value = value;
  }

  /**
   * Get the string representation of this direction.
   *
   * @return the direction string ("ltr", "rtl", or "auto")
   */
  public String getValue() {
    return value;
  }

  /**
   * Parse a text direction from string value.
   *
   * @param s the direction string ("ltr", "rtl", or "auto")
   * @return the {@code TextDirection}
   * @throws IllegalArgumentException if the string is null or unknown
   *
   * For example, {@code TextDirection.parse("ltr")} returns {@code TextDirection.LEFT_TO_RIGHT}.
   */
  public static TextDirection parse(String s) {
    if (s == null) {
      throw new IllegalArgumentException("TextDirection string cannot be null");
    }
    return switch (s.toLowerCase()) {
      case "ltr" -> LEFT_TO_RIGHT;
      case "rtl" -> RIGHT_TO_LEFT;
      case "auto" -> AUTO;
      default -> throw new IllegalArgumentException("Unknown TextDirection: " + s);
    };
  }

  @Override
  public String toString() {
    return value;
  }
}
